pkg/nodejs/prepare/application.go

```diff
@@ -91,7 +94,7 @@ http {
        root /u01/application/static;
 
        location /api {
-          proxy_pass http://localhost:9090;
+          proxy_pass http://${PROXY_PASS_HOST}:${PROXY_PASS_PORT};
```

Comment:
Her mener jeg man må sette noen headere slik at backenden klarer utlede urlen den ble kontaktet på skikkelig. Det er mulig man også må sette X-Forwarded-Proto, men den er litt mer klabbete dersom f.eks. bigip terminerer ssl, for deretter sende videre til routeren.

proxy_set_header Host $host;
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

Reply:
Disse headerne settes også av HAProxy.

For å det første Forwarded bør brukes av applikasjonen, ikke X-Forwarded-*[1]. Disse headerne settes av HAProxy, så jeg er litt skeptisk til å sette de på nytt.  I forslaget ditt over, vil Host settes til ip og port, som trolig ikke er det vi ønsker, X-Real-IP vil settes til IP til ruteren, som ikke er det vi ønsker, og X-Forwarded-For vil være uendret fra hva ruteren setter.

Jeg er i utgangspunktet enig i at de bør være der, problemet er at det krever en god del logikk for å få satt de riktig i alle situasjoner, så jeg heller mer mot å gjøre det enkelt, og heller sørge for at ruteren gjør ting riktig. Hva som er riktig, er også en diskusjon... Det er ikke gitt at hostheaderen bør skrives om.

Kan du komme på brukstilfeller hvor det er nødvendig å konktakte en web-app via en service, og ikke en ruter?

[1] https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Forwarded

Reply:
Jeg tror ikke det vil være relevant å kontakte en webapp gjennom en service. Kan ikke komme på noe use case for det, iallefall.

Jo mindre man trenger gjøre her, jo bedre; enig. Jeg har klabbet endel med disse headerene selv for å få dem riktig både for bruk fra localhost (utvikling) og bak Amazon ELB. Jeg har igrunnen ikke noen formening om hva som er riktig i alle tilfeller jeg heller, men det kan være vi bør teste de ulike scenariene (ikke alle som er like relevante) - iallefall ha dem i bakhodet;
- Lokal utvikling (gjennom nginx til lokal backend)
- Lokal utvikling (gjennom nginx til en backend på OpenShift (gjennom route))
- Deployet på OpenShift, backend gjennom service, aksessert gjennom route
- Deployet på OpenShift, backend gjennom service, aksessert gjennom route via webseal
- Deployet på OpenShift, backend gjennom service, aksessert gjennom route via bigip (denne fungerer ikke som tiltenkt?)

Ikke sikkert det er så relevant å bruke dette imaget til lokal utvikling, men hvis man ikke setter headerene eksplisitt vil nok ikke dét uansett fungere i forhold til serverside linkgenerering o.l.

Reply:
Jeg vet at disse funker:
* Lokal utvikling (gjennom nginx til en backend på OpenShift (gjennom route))
* Deployet på OpenShift, backend gjennom service, aksessert gjennom route
* Deployet på OpenShift, backend gjennom service, aksessert gjennom route via webseal

Denne vil funke gitt at man har ikke krever forwarded-headere (Spring mener jeg gjør dette riktig):
* Lokal utvikling (gjennom nginx til lokal backend)

Denne vil ikke fungere uansett hva vi gjør, siden ruteren setter forwarded-headere. Mulig det kan justeres med annoteringer:
* Deployet på OpenShift, backend gjennom service, aksessert gjennom route via bigip